crates/red_knot_python_semantic/resources/mdtest/narrow/truthiness.md

```diff
@@ -0,0 +1,189 @@
+# Narrowing For Truthiness Checks (`if x` or `if not x`)
+
+## Value Literals
+
+```py
+def bool_instance() -> bool:
+    return True
+
+def foo() -> Literal[0, -1, True, False, "", "foo", b"", b"bar"] | tuple[()] | None:
+    return 0
+
+x = foo()
+
+if x:
+    reveal_type(x)  # revealed: Literal[-1] | Literal[True] | Literal["foo"] | Literal[b"bar"]
+else:
+    reveal_type(x)  # revealed: Literal[0] | Literal[False] | Literal[""] | Literal[b""] | tuple[()] | None
+
+if not x:
+    reveal_type(x)  # revealed: Literal[0] | Literal[False] | Literal[""] | Literal[b""] | tuple[()] | None
+else:
+    reveal_type(x)  # revealed: Literal[-1] | Literal[True] | Literal["foo"] | Literal[b"bar"]
+
+if x and not x:
+    reveal_type(x)  # revealed: Never
+else:
+    reveal_type(x)  # revealed: Literal[-1, 0] | bool | Literal["", "foo"] | Literal[b"", b"bar"] | tuple[()] | None
+
+if not (x and not x):
+    reveal_type(x)  # revealed: Literal[-1, 0] | bool | Literal["", "foo"] | Literal[b"", b"bar"] | tuple[()] | None
+else:
+    reveal_type(x)  # revealed: Never
+
+if x or not x:
+    reveal_type(x)  # revealed: Literal[-1, 0] | bool | Literal["foo", ""] | Literal[b"bar", b""] | tuple[()] | None
+else:
+    reveal_type(x)  # revealed: Never
+
+if not (x or not x):
+    reveal_type(x)  # revealed: Never
+else:
+    reveal_type(x)  # revealed: Literal[-1, 0] | bool | Literal["foo", ""] | Literal[b"bar", b""] | tuple[()] | None
+
+if (isinstance(x, int) or isinstance(x, str)) and x:
+    reveal_type(x)  # revealed: Literal[-1] | Literal[True] | Literal["foo"]
+else:
+    reveal_type(x)  # revealed: Literal[b"", b"bar"] | tuple[()] | None | Literal[0] | Literal[False] | Literal[""]
+```
+
+## Function Literals
+
+Basically functions are always truthy.
+
+```py
+def flag() -> bool:
+    return True
+
+def foo(hello: int) -> bytes:
+    return b""
+
+x = flag if flag() else foo
+
+if x:
+    reveal_type(x)  # revealed: Literal[flag, foo]
+else:
+    reveal_type(x)  # revealed: Never
```

Comment:
I tried this approach, but I found that annotations don't work properly for FunctionLiteral.

For example, in the suggested code, the type of x is revealed as:
```
revealed type is `@Todo(Unsupported or invalid type in a type expression)`
```
Therefore, since the test is currently not working as expected, I reverted to using flag for now.

Reply:
Oh, of course, that was dumb of me :)